# Decision Tree Solution for Breast Cancer Dataset

In [1]:
import pandas as pd
import numpy as np
import random

## Loading Dataset into Pandas Dataframe

In [2]:
breast_cancer_df = pd.read_csv('data/data.csv', index_col='id')
breast_cancer_df

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  texture_worst  \
id                                            ...                  
842302                0.14710         0.2419  ...          17.33   
842517                0.07017         0.1812  ...          23.41   
84300903              0.12790         0.2069  ...          25.53   
84348301              0.10520         0.2597  ...          26.50   
84358402              0.10430         0.1809  ...          16.67   
...                       ...            ...  ...            ...   
926424                0.13890         0.1726  ...          26.40   
926682                0.09791         0.1752  ...          38.25   
926954                0.05302         0.1590  ...          34.12   
927241                0.15200         0.2397  ...          39.42   
92751                 0.00000         0.1587  ...          30.37   

          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
id                                                                           
842302             184.60      2019.0           0.16220            0.66560   
842517             158.80      1956.0           0.12380            0.18660   
84300903           152.50      1709.0           0.14440            0.42450   
84348301            98.87       567.7           0.20980            0.86630   
84358402           152.20      1575.0           0.13740            0.20500   
...                   ...         ...               ...                ...   
926424             166.10      2027.0           0.14100            0.21130   
926682             155.00      1731.0           0.11660            0.19220   
926954             126.70      1124.0           0.11390            0.30940   
927241             184.60      1821.0           0.16500            0.86810   
92751               59.16       268.6           0.08996            0.06444   

          concavity_worst  concave points_worst  symmetry_worst  \
id                                                                
842302             0.7119                0.2654          0.4601   
842517             0.2416                0.1860          0.2750   
84300903           0.4504       

In [3]:
# drop unnamed: 32 column
breast_cancer_df = breast_cancer_df.drop(['Unnamed: 32'], axis=1)
breast_cancer_df

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
id                                                                         
842302           M        17.99         10.38          122.80     1001.0   
842517           M        20.57         17.77          132.90     1326.0   
84300903         M        19.69         21.25          130.00     1203.0   
84348301         M        11.42         20.38           77.58      386.1   
84358402         M        20.29         14.34          135.10     1297.0   
...            ...          ...           ...             ...        ...   
926424           M        21.56         22.39          142.00     1479.0   
926682           M        20.13         28.25          131.20     1261.0   
926954           M        16.60         28.08          108.30      858.1   
927241           M        20.60         29.33          140.10     1265.0   
92751            B         7.76         24.54           47.92      181.0   

          smoothness_mean  compactness_mean  concavity_mean  \
id                                                            
842302            0.11840           0.27760         0.30010   
842517            0.08474           0.07864         0.08690   
84300903          0.10960           0.15990         0.19740   
84348301          0.14250           0.28390         0.24140   
84358402          0.10030           0.13280         0.19800   
...                   ...               ...             ...   
926424            0.11100           0.11590         0.24390   
926682            0.09780           0.10340         0.14400   
926954            0.08455           0.10230         0.09251   
927241            0.11780           0.27700         0.35140   
92751             0.05263           0.04362         0.00000   

          concave points_mean  symmetry_mean  ...  radius_worst  \
id                                            ...                 
842302                0.14710         0.2419  ...        25.380   
842517                0.07017         0.1812  ...        24.990   
84300903              0.12790         0.2069  ...        23.570   
84348301              0.10520         0.2597  ...        14.910   
84358402              0.10430         0.1809  ...        22.540   
...                       ...            ...  ...           ...   
926424                0.13890         0.1726  ...        25.450   
926682                0.09791         0.1752  ...        23.690   
926954                0.05302         0.1590  ...        18.980   
927241                0.15200         0.2397  ...        25.740   
92751                 0.00000         0.1587  ...         9.456   

          texture_worst  perimeter_worst  area_worst  smoothness_worst  \
id                                                                       
842302            17.33           184.60      2019.0           0.16220   
842517            23.41           158.80      1956.0           0.12380   
84300903          25.53           152.50      1709.0           0.14440   
84348301          26.50            98.87       567.7           0.20980   
84358402          16.67           152.20      1575.0           0.13740   
...                 ...              ...         ...               ...   
926424            26.40           166.10      2027.0           0.14100   
926682            38.25           155.00      1731.0           0.11660   
926954            34.12           126.70      1124.0           0.11390   
927241            39.42           184.60      1821.0           0.16500   
92751             30.37            59.16       268.6           0.08996   

          compactness_worst  concavity_worst  concave points_worst  \
id                                                                   
842302              0.66560           0.7119                0.2654   
842517              0.18660           0.2416                0.1860   
84300903            0.42450           0.4504                0.2430   
84348301       

In [4]:
# Check for NaN values
breast_cancer_df.isnull().sum()

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

### Implmenting Decision Tree Training Algorithm

In [5]:
class Tree:
    '''
    Class for decision tree classifier.
    '''
    class Node:
        '''
        Class for decision tree node.
        '''
        def __init__(self, feature, threshold, left, right):
            '''
            feature: feature to split on
            threshold: threshold to split on
            left: left subtree
            right: right subtree
            '''
            self.feature = feature
            self.threshold = threshold
            self.left = left
            self.right = right

        def __str__(self):
            return f'Tree.Node(feature={self.feature}, threshold={self.threshold}, \n\t , left={self.left},\n\t ,right={self.right}))'
        

    class Leaf:
        def __init__(self, label):
            '''
            label: label of the leaf node
            '''
            self.label = label

        def __str__(self):
            return str(self.label)

    def __init__(self, data, label_col, max_depth, criterion):
        '''
        data: pandas dataframe
        label_col: label column name
        max_depth: maximum depth of the tree
        criterion: splitting criterion
        '''
        self.data = data
        self.label_col = label_col
        self.max_depth = max_depth
        self.criterion = criterion
        self.root = self.fit()

        
    def majority_class(self, data):
        '''
        data: pandas dataframe
        return: majority class in the data
        '''
        return data[self.label_col].mode()[0]
    
    def feature_uncertainty_score(self, data):
        '''
        data: pandas dataframe
        return: uncertainty score of the data
        '''
        label_counts = data[self.label_col].value_counts()
        if self.criterion == 'gini':
            # gini impurity
            score = 1
            for label in label_counts:
                score -= (label/len(data))**2
            return score
        elif self.criterion == 'entropy':
            # entropy
            score = 0
            for label in label_counts:
                score -= (label/len(data)) * np.log2(label/len(data))
            return score
        elif self.criterion == 'misclassification':
            # misclassification error
            score = 1
            for label in label_counts:
                score = min(score, label/len(data))
            return score
        else:
            raise ValueError('Invalid criterion')
        
    def gain(self, data, feature, threshold):
        '''
        data: pandas dataframe
        feature: feature to split on
        threshold: threshold to split on
        return: information gain
        '''
        uncertainty = self.feature_uncertainty_score(data)
        left = data[data[feature] <= threshold]
        right = data[data[feature] > threshold]
        left_score = self.feature_uncertainty_score(left)
        right_score = self.feature_uncertainty_score(right)
        return uncertainty - (len(left)/len(data))*left_score - (len(right)/len(data))*right_score

    def best_split(self, data):
        '''
        data: pandas dataframe
        return: best feature and threshold to split on
        '''
        best_gain = -np.inf
        best_feature = None
        best_threshold = None
        for feature in data.columns:
            if feature != self.label_col:
                for threshold in data[feature]:
                    gain = self.gain(data, feature, threshold)
                    if gain > best_gain:
                        best_gain = gain
                        best_feature = feature
                        best_threshold = threshold
        return best_feature, best_threshold
    
    def fit(self):
        '''
        return: root node of the decision tree
        '''
        if self.max_depth == 0 or len(self.data) == 0:
            return self.Leaf(self.majority_class(self.data))
        feature, threshold = self.best_split(self.data)
        # print(feature, threshold)
        left = self.data[self.data[feature] <= threshold]
        right = self.data[self.data[feature] > threshold]
        left = left.drop([feature], axis=1)
        right = right.drop([feature], axis=1)

        if len(left) == 0 or len(right) == 0:
            return self.Leaf(self.majority_class(self.data))
        else:
            left = Tree(left, self.label_col, self.max_depth-1, self.criterion).root
            right = Tree(right, self.label_col, self.max_depth-1, self.criterion).root
            return self.Node(feature, threshold, left, right)
        
    def predict(self, data):
        '''
        data: pandas dataframe
        return: list of predictions
        '''
        predictions = []
        for index, row in data.iterrows():
            node = self.root
            while isinstance(node, self.Node):
                if row[node.feature] <= node.threshold:
                    node = node.left
                else:
                    node = node.right
            predictions.append(node.label)

        return predictions
    
    def __str__(self):
        return str(self.root)
    


### Spliting the dataset into training and testing data

In [6]:
# train test split 90% train 10% test
from sklearn.model_selection import train_test_split
data_train, data_test = train_test_split(breast_cancer_df, test_size=0.1, random_state=42)

### Training the Decision Tree Classifier

In [7]:
def best_depth(label_col, criterion, folds, max_depth=10):
    '''
    label_col: label column name
    criterion: splitting criterion
    folds: list of folds
    max_depth: maximum depth of the tree
    return: best depth and best accuracy
    '''
    best_depth = 0
    best_accuracy = 0
    
    for depth in range(max_depth):
        accuracy = 0
        
        for i in range(len(folds)):
            data_val = folds[i]
            data_train = pd.concat(folds[:i] + folds[i+1:])
            tree = Tree(data_train, label_col, depth, criterion)
            predictions = tree.predict(data_val.drop([label_col], axis=1))
            actual = data_val[label_col].tolist()
            accuracy += sum([1 for i in range(len(predictions)) if predictions[i] == actual[i]])/len(predictions)
        accuracy /= len(folds)
        print('Depth:', depth, 'Accuracy:', accuracy)
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_depth = depth
    return best_depth, best_accuracy

def best_criterion_depth(data: pd.DataFrame, label_col, k=5, max_depth=10):
    '''
    data: pandas dataframe
    label_col: label column name
    k: number of folds
    max_depth: maximum depth of the tree
    return: best criterion and best depth
    '''
    folds = []
    fold_size = len(data)//k
    for i in range(k):
        folds.append(data.iloc[i*fold_size:(i+1)*fold_size])
    best_criterion, best_criterion_best_depth = None, 0
    best_accuracy = 0

    for criterion in ['gini', 'entropy', 'misclassification']:
        print('Criterion:', criterion)
        criterion_best_depth, criterion_best_accuracy = best_depth(label_col, criterion, folds, max_depth)
        print('best depth:', criterion_best_depth, 'best accuracy:', criterion_best_accuracy)
        if criterion_best_accuracy > best_accuracy:
            best_accuracy = criterion_best_accuracy
            best_criterion = criterion
            best_criterion_best_depth = criterion_best_depth
    return best_criterion, best_criterion_best_depth
       


In [8]:
#best_criterion, best_criterion_best_depth = best_criterion_depth(data_train, 'diagnosis', k=5, max_depth=10)

Criterion: gini
Depth: 0 Accuracy: 0.6196078431372549
Depth: 1 Accuracy: 0.8745098039215685
Depth: 2 Accuracy: 0.9274509803921569
Depth: 3 Accuracy: 0.9294117647058824
Depth: 4 Accuracy: 0.9156862745098039
Depth: 5 Accuracy: 0.9274509803921569
Depth: 6 Accuracy: 0.9294117647058824
Depth: 7 Accuracy: 0.9274509803921569
Depth: 8 Accuracy: 0.9254901960784314
Depth: 9 Accuracy: 0.9254901960784314
best depth: 3 best accuracy: 0.9294117647058824
Criterion: entropy
Depth: 0 Accuracy: 0.6196078431372549
Depth: 1 Accuracy: 0.8745098039215685
Depth: 2 Accuracy: 0.9019607843137255
Depth: 3 Accuracy: 0.9313725490196079
Depth: 4 Accuracy: 0.9333333333333333
Depth: 5 Accuracy: 0.9392156862745098
Depth: 6 Accuracy: 0.9392156862745098
Depth: 7 Accuracy: 0.9372549019607843
Depth: 8 Accuracy: 0.9372549019607843
Depth: 9 Accuracy: 0.9372549019607843
best depth: 5 best accuracy: 0.9392156862745098
Criterion: misclassification
Depth: 0 Accuracy: 0.6196078431372549
Depth: 1 Accuracy: 0.8745098039215685
Dept

In [9]:
#best_criterion, best_criterion_best_depth

('entropy', 5)

The best criterion for the dataset and corresponding depth were found to be entropy and 5 respectively. This was done using cross validation, which was done iteratively for different depths and criterion. However, this approach may be inefficient, one way to improve this would be to use a grid search or use cross validation after implementing pruning to the decision tree so that we can train the tree to a greater depth and then prune it to the best depth.

In [11]:
tree = Tree(data_train, 'diagnosis', 5, 'entropy')
accuracy = sum([1 for i in range(len(data_test)) if tree.predict(data_test.drop(['diagnosis'], axis=1))[i] == data_test['diagnosis'].tolist()[i]])/len(data_test)
accuracy

0.9298245614035088

The accuracy of the trained classifier was found to be 92.98% on the test dataset.